### PR TITLE
Support for Tune search spaces

### PR DESCRIPTION
**Pull request by Yard1**
Adds support to Tune's search spaces, resolving https://github.com/ray-project/tune-sklearn/issues/124

### CODE REVIEW

**Comment by Yard1** [comment]
WIP. Will work more on it. No support for grid search yet (though do we need it?). Need to see if the tests go through. @krfricke FYI

**Comment by Yard1** [comment]
Also @richardliaw looks like CI is doing `pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.9.0.dev0-cp36-cp36m-manylinux1_x86_64.whl`. Shouldn't it be switched to ray-1.0.0?

**Comment by richardliaw** [comment]
hmm, can we update it to 1.1.0.dev0? (see the wheel links here https://docs.ray.io/en/latest/installation.html)

**Comment by Yard1** [comment]
@krfricke How are tune search spaces handled in grid search? The only thing left for this PR is determining how to handle grid searching. I think that it would be fine to leave it as-is (that is only support a dict of lists instead of any search spaces), but I am open to suggestions.

**Comment by richardliaw** [comment]
@Yard1 feel free to ping Kai again, perhaps on slack, after 48 hours :)

**Comment by krfricke** [comment]
In Tune, grid search is represented by a `Categorical` domain with a `Grid` sampler. You can also use the [`Domain.is_grid()`](https://github.com/ray-project/ray/blob/master/python/ray/tune/sample.py#L50) method to determine if a domain should be grid searched. 

In Tune, grid search means that all combinations of all grid search items (if there is more than one grid search item) should be generated at least once. E.g. with `"a": tune.grid_search([1, 2])` the basic variant generator will generate two trials, one with `a=1` and one with `a=2`. The `num_samples` parameter will multiply this number, i.e. if `num_samples=2`, _four_ trials will be generated here. Does this help?

**Comment by Yard1** [comment]
@krfricke So how would a user pass search spaces to be grid searched?

**Comment by krfricke** [comment]
So in regular Ray Tune a user would pass something like

```
config = {
    "var": tune.grid_search([1, 2, 3]),
    "rnd": tune.uniform(0, 5)
}
tune.run(
    trainable,
    config=config)
```

With `num_samples=1` this would create three trials, one for each grid serach value (1, 2 and 3). The `rnd` value is drawn randomly for each of these trials.

**Comment by Yard1** [comment]
Great, thanks. This clears it up.

**Comment by Yard1** [comment]
@richardliaw @krfricke Tests are passing, Mac fails due to an outdated Ray version, I think. I'll double check, but in the meantime, it's ready for review.

It's also not working with the currently released Ray 1.0.0 as it's missing the functionality @krfricke added later.

**Comment by richardliaw** [comment]
1.0.1 is out now :)

**Comment by Yard1** [comment]
@richardliaw Great to hear that! I need to put on notifications for the main repo, hah

File: tune_sklearn/tune_gridsearch.py [hunk 1]
```diff
@@ -176,9 +178,37 @@ def __init__(self,
...
+                        "Parameter grid for parameter ({0}) needs to"
+                        " be a tune.grid_search, list or numpy array,"
+                        " but got ({1})."
+                        " Single values need to be wrapped in a list"
+                        " with one element.".format(name, type(v)))
+
+                if len(v) == 0:
+                    raise ValueError(
+                        "Parameter values for parameter ({0}) need "
+                        "to be a non-empty sequence.".format(name))
```
Comment (richardliaw) [inline tune_sklearn/tune_gridsearch.py:210]: can this be a helper function (not scoped within the class?)

**Comment by richardliaw** [comment]
This looks good! though:
* it seems to be failing on mac build?
* can we update the readme/docstrings?

Also, @krfricke please take a look!

File: tune_sklearn/tune_gridsearch.py (hunk 1 above)
Comment (Yard1) [reply tune_sklearn/tune_gridsearch.py]: Done, moved to utils

**Comment by Yard1** [comment]
@richardliaw Updated docstrings and readme. Not sure what's up with Mac, I'll look into it tomorrow. Maybe it's downloading the wrong version?

**Comment by Yard1** [comment]
@krfricke Mac CI fails on `test_timeout`, could you take a look?

**Comment by krfricke** [comment]
Yep, I'll look into it tomorrow

**Comment by krfricke** [comment]
Seems there's some more overhead for the test_timeout (18.47 seconds instead of 18). Without the time budget the test would run for 50+ seconds. Should we just raise that limit to 25 to account for environment/MacOS overhead? Do you just want to include it in this PR since it is a 1 line change?

**Comment by Yard1** [comment]
@krfricke Yeah, sure. Do you want to make the change or should I?

**Comment by richardliaw** [comment]
Antoni, probably faster for yourself to change it ;)

**Comment by Yard1** [comment]
@richardliaw Touché!

**Comment by Yard1** [comment]
@richardliaw Mac now errors instead of failing. It was doing that with other PRs - I think it's something to do with running out of memory? I think it should be good to merge though, considering Linux passes.

File: tests/test_gridsearch.py [hunk 2]
```diff
@@ -702,6 +705,18 @@ def test_local_mode(self):
...
+        clf = MockClassifier()
+        grid_search = TuneGridSearchCV(
+            clf, {
+                "foo_param": tune.grid_search([1, 2, 3]),
+                "bar_param": [1, 2]
+            },
+            refit=False,
+            cv=3)
+        grid_search.fit(X, y)
+
```
Comment (krfricke) [inline tests/test_gridsearch.py:719]: Should this test if all combinations were seen?

**Review by krfricke — APPROVED** [review]
Looks good, maybe check for all grid search combinations in test.

File: tests/test_gridsearch.py (hunk 2 above)
Comment (Yard1) [reply tests/test_gridsearch.py:725]: Good call, added

**Comment by richardliaw** [comment]
hmmm it looks like Mac tests are failing but it's not an out of memory issue?

**Comment by Yard1** [comment]
@richardliaw Looks like the functionality needed for this PR is not yet in release Ray - that's what Mac uses, and it's what the exception on it is related to.

**Comment by richardliaw** [comment]
Got it - a couple things

1. We can introduce new features (without regressions) not compatible with released Ray, but we must make sure the user knows what is the error.
2. We should avoid introducing a failing test into CI; it should be marked as skipped if anything.

Thanks!

**Comment by Yard1** [comment]
@richardliaw Got it. I think we could just wait a while on this PR, unless we are planning to release a new version of tune-sklearn soon.

**Pull request merged**